OBSDOCS-1976: Document support for GCP, Azure, and AWS STP tokens (#94801)
openshift/openshift-docs: modules/distr-tracing-tempo-object-storage-setup-aws-sts-cco-install.adoc

// Module included in the following assemblies:
//
//* observability/distr_tracing/distr-tracing-tempo-installing.adoc

:_mod-docs-content-type: PROCEDURE
[id="setting-up-amazon-s3-storage-with-security-token-service_{context}"]
= Setting up the Amazon S3 storage with the Security Token Service

You can set up the Amazon S3 storage with the Security Token Service (STS) by using the AWS Command Line Interface (AWS CLI).

:FeatureName: The Amazon S3 storage with the Security Token Service
include::snippets/technology-preview.adoc[leveloffset=+1]

.Prerequisites

* You have installed the latest version of the AWS CLI.
* You have installed and configured the Cloud Credential Operator in your cluster.

.Procedure

. Create an AWS S3 bucket.

. Create the following `trust.json` file for the AWS IAM policy that will set up a trust relationship for the AWS IAM role, created in the next step, with the service account of the TempoStack instance:
+
[source,yaml]
----
{
    "Version": "2012-10-17",
    "Statement": [
      {
        "Effect": "Allow",
        "Principal": {

>           "Federated": "arn:aws:iam::${<aws_account_id>}:oidc-provider/${<oidc_provider>}" # <1>

Author: Should I rewrite this as follows?
```suggestion
          "Federated": "arn:aws:iam::<aws_account_id>:oidc-provider/<oidc_provider>" # <1>
```

        },
        "Action": "sts:AssumeRoleWithWebIdentity",
        "Condition": {
          "StringEquals": {

>             "${OIDC_PROVIDER}:sub": [

Author: Should I rewrite this as follows?
```suggestion
            "<oidc_provider>:sub": [
```

Reviewer: yes!

----

. Optional: If you are not using the CCO, skip this step. If you are using the CCO, configure the cloud provider environment for the {TempoOperator}:
+
[source,terminal]
----
$ oc patch subscription <tempo_operator_sub> \ # <1>
          -n <tempo_operator_namespace> \ # <2>
          --type='merge' -p '{"spec": {"config": {"env": [{"name": "ROLEARN", "value": "'"<role_arn>"'"}]}}}' # <3>
----
<1> The name of the {TempoOperator} subscription.
<2> The namespace of the {TempoOperator}.
<3> The AWS STS requires adding the `ROLEARN` environment variable to the {TempoOperator} subcription. As the `<ROLE_ARN>` value, add the Amazon Resource Name (ARN) of the AWS IAM role that you created in step 3.

. In the {product-title}, create an object storage secret with keys as follows:
+
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: <name>
stringData:
  bucket: <s3_bucket_name>
  region: <s3_region>
  role_arn: <s3_role_arn>
type: Opaque
----

. When the object storage secret is created, update the custom resource of the {TempoShortName} instance:

** An example for the `TempoStack` custom resource:
+
[source,yaml]

> ----

Author: :slightly_smiling_face: Thank you.
I rewrote line 108 as a block title syntax.
The user-replaced values are clear about what is required. I'd like to avoid fluff and distraction by repeating the user-replaced values as footnotes.

apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: <name>
  namespace: <tempo_ns>
spec:
  storage:
    secret:
      name: aws-sts
      type: s3
      credentialMode: token-cco # <1>
----
<1> Optional: If you are not using the CCO, omit this line. If you are using the CCO, add this parameter with the `token-cco` value.

** An example for the `TempoMonolithic` custom resource:
+

> [source,yaml]

Author: :slightly_smiling_face: Thank you.
I rewrote line 126 as a block title syntax.
The user-replaced values are clear about what is required. I'd like to avoid fluff and distraction by repeating the user-replaced values as footnotes.

----

. Configure the cloud provider environment for the {TempoOperator}:
+
[source,terminal]
----
$ oc patch subscription <tempo_operator_sub> \ # <1>
          -n <tempo_operator_namespace> \ # <2>
          --type='merge' -p '{"spec": {"config": {"env": [{"name": "ROLEARN", "value": "'"<role_arn>"'"}]}}}' # <3>
----
<1> Operator Subscription Name.
<2> Namespace of the {TempoOperator}.
<3> The AWS STS requires adding the `ROLEARN` environment variable to the Operator subcription. As the `<ROLE_ARN>` value, add the Amazon Resource Name (ARN) of the AWS IAM role that you created in step 3.

. In the {product-title}, create an object storage secret with keys as follows:
+
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: <name>
stringData:
  bucket: <s3_bucket_name>
  region: <s3_region>
  role_arn: <s3_role_arn>
type: Opaque
----

. When the object storage secret is created, add the `credentialMode` parameter to the `TempoStack` custom resource:
+
[source,yaml]
----
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: <name>
  namespace: <tempo_ns>
spec:
  storage:
    secret:
      name: aws-sts
      type: s3 # <1>

>       credentialMode: token-cco # <2>

Author: @rubenvp8510, can I add a comment that this line is `Optional`?

Reviewer: For CCO is not optional, usually if you don't specify it we can infer it from secret content, but not for CCO

Reviewer: ha yes, yes, if you don't want to use CCO you can configure using other methods.

Reviewer 2: The credentialMode needs to be set to token-cco only when using CCO else we can skip the parameter or set it to token for STS without CCO.

----
<1> Currently, this is the only backend supported togheter with the CCO.

> <2> Must be set to the `token-cco` value.

Author: @rubenvp8510 WDYT? Like this:
```suggestion
<2> Optional: If you are using the CCO, add this parameter with the `token-cco` value.
```

Reviewer: Yes this is good :)
